Abstract record interface

mitchcapper opened the pull request:
This primarily exposes two new interfaces for files and directories:
```csharp
public interface IAbstractRecord {
		DateTime CreationTimeUtc { get; }
		FileAttributes FileAttributes { get; }
		string FileName { get; }
		bool IsDirectory { get; }
		bool IsSymlink { get; }
		DateTime LastAccessTimeUtc { get; }
		DateTime LastWriteTimeUtc { get; }
		long FileId { get; }
		long FileSize { get; }
		Streams.IBuffer FileContent { get; }
		VfsDirEntry GetAsDirEntry();
		IVfsFile GetAsFile();
		IAbstractDirectory GetAsAbstractDirectory();
}

public interface IAbstractDirectory : IAbstractRecord {
		IEnumerable<IAbstractRecord> AllEntries { get; }
}
```
and two methods on file systems:
```csharp
Vfs.IAbstractRecord GetAbstractRecord(string path);
string GetSymlinkTarget(Vfs.IAbstractRecord dirEntry);
```
First, thanks for continuing to develop such an awesome library :)  Secondly, while this PR I think brings some improvements (although at the cost of an additional interface) they are pretty minor at only 15% (given most wait is blocking IO that just gets cached in the current).  If this isn't desired it may be worth at least adding IsDirectory and IsSymlink to the requirements on DiscFile entries so those two bytes avoid string lookups.


My current primary use is to index my NAS that contains 20+ years of backups in all forms and formats just giving me a file name index of the files in it (and contained disk images and archives).  For DiscUtils if I find something it can open I just traverse from the root recursively through the drive to get every file/directory.  I have some sample code at the bottom.

My issue happened when I went to log the symlink target and noticed there wasn't a good way to do so. DiscFileSystemInfo doesn't provide a way to get the symlink target and neither does DiscFileSystem.  The clear choice was to cast to the actual concrete type and assume it has some way to do so, but before i went to try and condition switch on every possible format I looked further.  I started to run into the issues I documented in #39.

Vfs kind of looks like it does and seems to be used by almost everything except NTFS for some reason.  In addition there is no way to actually access VfsFileSystem as it primarily seems an internal only target.  You can get the VfsFileSystemFacade through casting for some file systems but not to the underlying wrapped item.

Next, as I started tracing how things worked I noticed there seemed to be some perceived inefficiency, at least in how I was use the library.  For each DiscFileInfo file I was calling .Attributes to check if it was a symlink (reparse point) but this did a full re-lookup of the file info based on its string path, a few conversions to File class and such even though the information was already had in the DirEntry we originally accessed.  I do see that many things were cached (although still the string file path was used as a key every time for many calls even though I was operating on a DiscFileInfo object).  I was actually surprised how fast it runs as it is very speedy at iterating systems.   I know you folks have used DiscUtils to mount images as live drives in Windows so I was curious to see how fast it was (although it could also use a severely different set of code paths potentially?) but I didn't have easy luck doing so.  I don't think your new image mounter: https://github.com/ArsenalRecon/Arsenal-Image-Mounter has an option in the free version to use DiscUtils (or the powershell / cli).  I think DiscUtilsDevio is still around for windows but uses an older copy of DiscUtils so my changes wouldn't have any effects there.    

Clearly though as people use it professionally it must be fast, but I wasn't seeing how to do that for my use case.

Still I figured it might be helpful to get a common interface for files that would work across file systems and support things like symlink resolution, directly knowing if something a symlink or a directory or a file,  and access to a "FileID".  Given both Windows and most *nix systems support a drive specific un-changing file ID I figured it would be helpful to expose.   For one it could avoid a recursion bug if you were to cache the ID's you already visited.  At the same time improve performance by allowing the user to have more "low level access" but still abstracted but able to get a folder and all its records.  I am bad at naming things so try to ignore that:)  This is also a bit hacky for one, I stuck this in Vfs but, NTFS doesn't use that so I still need to add it there.  There is also a real trashy hack around the DriveRoot as I need a DirEntry for my current implementation so in VfsFileSystem you will find "FakeRootDirEntry".  For one, this will return the wrong inode value for ext.

100% a draft but I figured it is at least a working example to get some feedback on in terms of interest.

Anyway at the end of the day, while I think it is more performant on a quick image test the improvements were pretty minor than I expected.  1m42s for the original traversing vs 1m27s for this code path.  It is only 15% faster.   Granted I did have to hard code disable symlink resolution on the old codepath to get it to complete but I doubt that really made a huge difference.
If user time is primarily the code itself though vs the system time a bit larger of a gap:
user    0m17.890s
sys     1m24.216s

user    0m4.181s
sys     1m23.801s




Here is my old style:

```csharp

int ExploreDir(DiscFileSystem fileSystem, string basePath, string currentPath) {
var dirInfo = fileSystem.GetDirectoryInfo(currentPath);
curDirLastWrite = dirInfo.LastWriteTimeUtc;
var directories = dirInfo.GetDirectories();
foreach (DiscDirectoryInfo directory in directories) {

				if (directory.Attributes.HasFlag(FileAttributes.ReparsePoint)) {
					callbacks.PrintSymLink(basePath + "/" + directory.FullName.Replace('\\', '/').TrimStart('/'),
						fileSystem.GetSymlinkTarget(fileSystem.GetAbstractRecord(directory.FullName)), directory.LastWriteTime);
					continue;
				}
				dirTotal += ExploreDir(fileSystem, basePath, directory.FullName);

}

var files = dirInfo.GetFiles();
foreach (var file in files) {
				var fullPath = basePath + "/" + file.FullName.Replace('\\', '/').TrimStart('/');
				if ((file.Attributes & FileAttributes.ReparsePoint) != 0) 
					callbacks.PrintSymLink(fullPath, fileSystem.GetSymlinkTarget(fileSystem.GetAbstractRecord(file.FullName)), file.LastWriteTime);
				else {
					callbacks.PrintItem(fullPath, file.LastWriteTime, file.Length);
					dirTotal += file.Length;
				}
}
}
```


Here the newer style:

```csharp
private long ExploreDir(DiscFileSystem fileSystem, string basePath, string currentPath,  IAbstractDirectory dir = null) 
var curDir = dir ?? fileSystem.GetAbstractRecord(currentPath) as IAbstractDirectory;
curDirLastWrite = curDir.LastWriteTimeUtc;
foreach (var record in curDir.AllEntries) {
				var recordFullPath = Path.Join(currentPath, record.FileName).Replace('\\', '/').TrimStart('/');
				if (record.IsSymlink)
					callbacks.PrintSymLink(basePath + "/" + recordFullPath, fileSystem.GetSymlinkTarget(record).Replace('\\', '/'), record.LastWriteTimeUtc);
				else if (record.IsDirectory) {
					dirTotal += PrintFolder(fileSystem, basePath, recordFullPath, record.GetAsAbstractDirectory());
				} else {
					var fullPath = basePath + "/" + recordFullPath.TrimStart('/');
					callbacks.PrintItem(fullPath, record.LastWriteTimeUtc, record.FileSize);
					dirTotal += record.FileSize;

				}
```




Finally (although unrelated), minor item, but it may be worth improving the generic VirtualDisk.OpenDisk docs about the fact passing it a file will only use the extension to determine possible compat decoders.   I maintain a list of my extensions that might be raw drives (say .bin or .iso if I was lazy) and then just call it with the forced "RAW" type if they fail to open automatically.
```csharp
disk = VirtualDisk.OpenDisk(file, FileAccess.Read);
if (disk.Partitions == null || disk.Partitions.Count == 0) {
				if (MayBeRawExtensions.Any(a => file.EndsWith(a, StringComparison.OrdinalIgnoreCase))) {
					disk.Dispose();
					disk = VirtualDisk.OpenDisk(file, "RAW", FileAccess.Read, default, default);
				}
}

mitchcapper committed:
Initial implementation hooked into Vfs for now

Olof-Lagerkvist commented:
> First, thanks for continuing to develop such an awesome library :) Secondly, while this PR I think brings some improvements (although at the cost of an additional interface) they are pretty minor at only 15% (given most wait is blocking IO that just gets cached in the current). If this isn't desired it may be worth at least adding IsDirectory and IsSymlink to the requirements on DiscFile entries so those two bytes avoid string lookups.

Thanks a lot for your contributions! My goal with my optimizations so far has mainly been to reduce Garbage Collector pressure by avoiding a lot of unnecessary heap objects. Also, to some extent, added cache logic to some frequently accessed code paths in file system implementations. But there are still many places this could be optimized further and there are also lots of other things that could be optimized in various ways. I really appreciate your ideas about this!

> Vfs kind of looks like it does and seems to be used by almost everything except NTFS for some reason. In addition there is no way to actually access VfsFileSystem as it primarily seems an internal only target. You can get the VfsFileSystemFacade through casting for some file systems but not to the underlying wrapped item.

`VfsFileSystem` is an internal abstraction logic for some file systems with read-only implementations that have a lot in common. File systems like NTFS and FAT are a bit different in many aspects and we also provide write support for them as well as access to internal meta data to some extent.

> Next, as I started tracing how things worked I noticed there seemed to be some perceived inefficiency, at least in how I was use the library. For each DiscFileInfo file I was calling .Attributes to check if it was a symlink (reparse point) but this did a full re-lookup of the file info based on its string path, a few conversions to File class and such even though the information was already had in the DirEntry we originally accessed. I do see that many things were cached (although still the string file path was used as a key every time for many calls even though I was operating on a DiscFileInfo object). I was actually surprised how fast it runs as it is very speedy at iterating systems. I know you folks have used DiscUtils to mount images as live drives in Windows so I was curious to see how fast it was (although it could also use a severely different set of code paths potentially?) but I didn't have easy luck doing so. I don't think your new image mounter: https://github.com/ArsenalRecon/Arsenal-Image-Mounter has an option in the free version to use DiscUtils (or the powershell / cli). I think DiscUtilsDevio is still around for windows but uses an older copy of DiscUtils so my changes wouldn't have any effects there.

`DiscUtilsDevio` did never support file systems in that way. It only used DiscUtils disk image implementations such as vhd, vmdk etc to emulate disks. File systems were mounted by Windows file system drivers. This logic has since been moved to Arsenal Image Mounter command line tool `aim_cli`. But in either case, they do not have the features you are looking for.

I have a project called `DiscUtilsFs` that uses DiscUtils file system implementations to mount file systems in a disk image through `DiscUtils.MountDokan` and Dokan driver in Windows and `DiscUtils.MountFuse` and Fuse driver on Linux and FreeBSD. That is probably more what you are looking for here.
https://github.com/LTRData/DiscUtilsFs/

> Clearly though as people use it professionally it must be fast, but I wasn't seeing how to do that for my use case.

The logic that exposes an `IFileSystem` from DiscUtils as a Dokan file system is here:
https://github.com/LTRData/DiscUtils/blob/LTRData.DiscUtils-initial/Library/DiscUtils.MountDokan/DokanDiscUtils.cs

It does not use objects like `DiscFileSystemInfo` at all.

> Still I figured it might be helpful to get a common interface for files that would work across file systems and support things like symlink resolution, directly knowing if something a symlink or a directory or a file, and access to a "FileID". Given both Windows and most *nix systems support a drive specific un-changing file ID I figured it would be helpful to expose. For one it could avoid a recursion bug if you were to cache the ID's you already visited. At the same time improve performance by allowing the user to have more "low level access" but still abstracted but able to get a folder and all its records. I am bad at naming things so try to ignore that:) This is also a bit hacky for one, I stuck this in Vfs but, NTFS doesn't use that so I still need to add it there. There is also a real trashy hack around the DriveRoot as I need a DirEntry for my current implementation so in VfsFileSystem you will find "FakeRootDirEntry". For one, this will return the wrong inode value for ext.
> 
> 100% a draft but I figured it is at least a working example to get some feedback on in terms of interest.

This is a really interesting idea. I have had ideas for exposing more meta information for `VfsFileSystem` style file systems before. One idea I had was to introduce an option in `DiscFileSystemOptions` to turn off resolving symlinks to allow the actual symlinks to be opened and queried by some other methods. Another idea to have a flag in `GetFiles`, `OpenFile` etc to list files without resolving symlinks and to open symlinks directly there.

mitchcapper committed:
WinFS and NTFS abstract examples

mitchcapper commented:
> My goal with my optimizations so far has mainly been to reduce Garbage Collector pressure by avoiding a lot of unnecessary heap objects

I realize that this certainly adds more objects if utilizing all the current functionality the main goal was to reduce total allocations for many use cases that might include traversing or looking through the filesystem without accessing most of the files themselves. I totally understand though if this isn't the way you end up going, but I think a common interface that works across all would be beneficial to users.

> I have a project called DiscUtilsFs that uses DiscUtils file system implementations to mount file systems in a disk image through DiscUtils.MountDokan and Dokan driver in Windows and DiscUtils.MountFuse and Fuse driver on Linux and FreeBSD. That is probably more what you are looking for here.

Fun! and had not heard of Dokan before seems like a great way to go.   I only wanted to test it for seeing the speed but useful all the same.

> This is a really interesting idea. I have had ideas for exposing more meta information for VfsFileSystem style file systems before.

I think it is important this abstraction works for NTFS as well.  NTFS has all the same fields IAbstractRecord exposes and that would give one unified interface.   I added the two current abstraction access points to IFileSystem with the idea that if something like this was picked up it would work across everything for a standardized access method.

> One idea I had was to introduce an option in DiscFileSystemOptions to turn off resolving symlinks to allow the actual symlinks

I think an option in there makes a lot of sense.  

> Another idea to have a flag in GetFiles, OpenFile etc to list files without resolving symlinks and to open symlinks directly there.

I could see it maybe on GetFiles as an option although OpenFile I am not sure it makes sense on (as opening the stream to the target should always be the goal?).  If someone doesn't want OpenFile to follow a symlink they could check it before hand.  One option would be to have the DiscFileSystemOptions have an option like "returnSymlinksAsTargetByDefault".  and then a Bool? on GetFiles allowing that behavior to be overwritten.

I updated this PR to include NTFS and WIM image support to show what that could look like.  In the process I didn't see any reparse resolution so added that for the two common windows formats.  There is some discrepancy as IsSymLink returns true if there is any reparse point present even though not all Reparse Points can be represented simple as links.  This was done as it makes the IsSymLink call far cheaper (no need to actually read the reparse point and check the tags) and I would think for most cases gives a result that is fairly reasonable.


One odd thing, maybe I made a mistake or maybe it is a WIM handling issue.   I created a small WIM image to test with and I noticed for one of the directory links when I call GetEntry(path); on WimFileSystem.cs the resulting entry had an all zero hash for some reason.  This only occurred on one symlink and not any of the others.  This then causes it to fail when you get the reparse point.   I found though if I find it returns all zeros I can call GetFileHash and then it has the correct hash.   In the attached example WIM file it is the "photos/test.txt" path that has the problem.

Here is the updated with the hash correction fix:

```csharp
   public ReparsePoint GetReparsePoint(string path)
    {
        var dirEntry = GetEntry(path);

        var hdr = _file.LocateResource(dirEntry.Hash)
		var hash = dirEntry.Hash;
		if (Utilities.IsAllZeros(hash))
			hash = GetFileHash(path);
        var hdr = _file.LocateResource(hash)
            ?? throw new IOException("No reparse point");

        using var s = _file.OpenResourceStream(hdr);
        var buffer = new byte[s.Length];
        s.ReadExactly(buffer, 0, buffer.Length);
        return new ReparsePoint((int)dirEntry.ReparseTag, buffer);
    }
```


For the most part these are lighter weight than the normal enumeration classes but with NTFS there is a bit of an oddity.  For the others the concrete classes that implement IAbstractDirectory (NtfsAbstractDirectory for ntfs) inherit from whatever class implemented IAbstractRecord (NtfsAbstractRecord).  With NTFS this is also true, but the NtfsAbstractDirectory has a much higher cost to building building vs NtfsAbstractRecord which is very light. NtfsAbstractDirectory requires the entire Directory/File object from DiscUtils.NTFS to be created.   As such all results from a directory.AllEntries() call returns only true IAbstractRecords and only if record.GetAsAbstractDirectory() is called does it result in a DiscUtils.NTFS.Directory being created.  In all other classes while they have GetAsAbstractDirectory() you can just cast IAbstractFile to IAbstractDirectory if it is a directory.


I noticed as well there is no WIM factory, meaning the style of code:

```csharp
disk = VirtualDisk.OpenDisk(file, FileAccess.Read);
var volMgr = new VolumeManager();
volMgr.AddDisk(disk);
var vols = volMgr.GetLogicalVolumes();
foreach (var vol in vols) {
	var fsses = FileSystemManager.DetectFileSystems(vol);
	using var fileSystem = fsses.Single().Open(vol);
```

doesn't work for WIM images.  This was easy enough to work around with another special case handler like:
```
if (file.EndsWith(".wim", StringComparison.OrdinalIgnoreCase)) {
				var wimFile = new WimFile(File.OpenRead(file));
				for (var x = 0; x < wimFile.ImageCount; x++) {
					var img = wimFile.GetImage(x);
					var partitionPath = imagePath + "/p" + x;
					if (!img.Root.Exists)
						continue;
					PrintFolder(img, partitionPath);
				}
```
but I am not sure if a fake factory wouldn't be beneficial for general users.  I realize WIM is something of an odd format given it supports multiple partitions/images while also being a custom file format itself.
[ntfs_sym_final.zip](https://github.com/user-attachments/files/22353257/ntfs_sym_final.zip)

Olof-Lagerkvist commented:
The way I see it, we could add a new interface like `ISymLinksFileSystem` that derives from `IFileSystem` and just provides a few more features. Much like how interfaces like `IClusterBasedFileSystem` work. It could provide features for enumerating directories and returning symlink information instead of targets and methods for querying symlink targets by path etc. Both `VfsFileSystem` derived file systems and NTFS could implement this interface, while file systems such as FAT could be left without this implementation.

mitchcapper commented:
Agreed.  I attempted to expose a FileId which is generally already had based on the cache id but one could add an interface for that.  I figure it would be a good way to avoid circular loops in a traversal.

My additional code path tries to avoid creating File objects (and when possible Directory objects) that are specific to that filesystem so that accessing metadata is cheaper.  The normal path instead creates it for everything and then uses caching with path based lookup for any data access so accessing the last modified time, Attributes, and size results in 3x traversing the path  to get to the directory entry then the File record is created (or retrieved from the cache) and then the property accessed.   Still, as mentioned, the current code is quite performant and there is little gained by the new interface otherwise. 

Anyway anything worth while feel free to strip out at least the reparse code should be useful:)

mitchcapper committed:
Finished most others minus nativeFS didn't add native calls to get inode/fileid

mitchcapper committed:
Fixed Fat abstract directory support.

mitchcapper committed:
NTFSish system improvements to symlink/reparse points

IAbstractRecord Filename standardization for NTFSish as well

mitchcapper committed:
Proper conversion for abstract entry to directory
